contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,404 @@
+
+  // Last error encountered during attach operation, if any.
+  // +optional
+  // This field must only be set by the entity completing the attach
+  // operation, i.e. the external-attacher.
+  AttachError VolumeError `json:"attachError,omitempty" protobuf:"bytes,3,opt,name=attachError,casttype=VolumeError"`
+
+  // Last error encountered during detach operation, if any.
+  // +optional
+  // This field must only be set by the entity completing the detach
+  // operation, i.e. the external-attacher.
+  DetachError VolumeError `json:"detachError,omitempty" protobuf:"bytes,3,opt,name=detachError,casttype=VolumeError"`
+}
+
+type VolumeError struct {
+  // Time the error was encountered.
+  // +optional
+  Time metav1.Time `json:"time,omitempty" protobuf:"bytes,1,opt,name=time"`
+
+  // String capturing the error encountered during Attach operation.
+  // This string maybe logged, so it should not contain sensitive
+  // information. 
+  // +optional
+  Message string `json:",message,omitempty" protobuf:"bytes,2,opt,name=message"`
+}
+```
+
+### Kubernetes In-Tree CSI Volume Plugin
+
+A new in-tree Kubernetes CSI Volume plugin will contain all the logic required for Kubernetes to communicate with an arbitrary, out-of-tree, third-party CSI compatible volume driver.
+
+The existing Kubernetes volume components (attach/detach controller, PVC/PV controller, Kubelet volume manager) will handle the lifecycle of the CSI volume plugin operations (everything from triggering volume provisioning/deleting, attaching/detaching, and mounting/unmounting) just as they do for existing in-tree volume plugins.
+
+#### Proposed API
+
+A new `CSIVolumeSource` object will be added to the Kubernetes API. It will be part of the existing `VolumeSource` and `PersistentVolumeSource` objects. This will enable Kubernetes users to use the new volume just like existing volume via a `PersistentVolume` or as a direct reference in a pod.
+
+```GO
+type CSIVolumeSource struct {
+  // Driver is the name of the driver to use for this volume.
+  // Required.
+  Driver string `json:"driver" protobuf:"bytes,1,opt,name=driver"`
+
+  // VolumeHandle is the unique volume name returned by the CSI volume
+  // plugin’s CreateVolume to refer to the volume on all subsequent calls.
+  VolumeHandle string `json:"volumeHandle" protobuf:"bytes,2,opt,name=volumeHandle"`
+
+  // Optional: SecretRef is reference to the secret object containing
+  // sensitive information to pass to the CSI driver during
+  // ControllerPublish and NodePublish. This may be empty if no secret
+  // object is specified. If the secret object contains more than one
+  // secret, all secrets are passed.
+  // This is a local object reference so that only secrets within the
+  // same namespace as the pod referencing this volume can be referenced.
+  // +optional
+  SecretRef *LocalObjectReference `json:"secretRef,omitempty" protobuf:"bytes,3,opt,name=secretRef"`
```

Comment:
you cannot limit to a specific secret type. you can limit to a specific namespace. I'd expect the following:
* external provisioner is granted read/write access to secrets in a specific namespace (e.g. `myfs-provisioner`)
* storage class for that provisioner sets the config parameter controller the location of secrets to that namespace `,"secretNamespace":"myfs-provisioner",...`
* the provisioner creates persistent volumes referencing secrets in that namespace
* if it provisions secrets, it provisons them in that namespace

Reply:
@liggitt yes, this works for provisioners! And it works with the proposal in https://github.com/kubernetes/community/pull/1258#discussion_r146369401, which adds two additional sets of secrets: one for attaching and one for mounting.

For the attaching secrets, the same thing you outlined here should work: i.e. a `my-external-attacher` namespace, and limiting the external-attacher's access to just that namespace.

For the mounting secrets, which must live in the user's namespace, since those will be dereferenced by kubelet, kubelet already has access to pod secrets.

Reply:
> For the mounting secrets, which must live in the user's namespace

Some mount secrets are not appropriate for namespaced users to have access to.

> since those will be dereferenced by kubelet, kubelet already has access to pod secrets.

as PV API objects are being refactored to allow specifying namespaces of secrets, the node authorizer automatically allows kubelets to access those secrets needed for PVs used by its pods

Reply:
if this is going to be added to pods and pv's, we need the same split here as is being added for the other PV sources... restricting to the local namespace when used in a Pod, allowing specifying a secret namespace when used in a PV

Reply:
> Some mount secrets are not appropriate for namespaced users to have access to.

I think this is worth clarifying if this will still be true with CSI. Mount should only require end user credentials. Any credentials required to access the storage control plane must be injected directly into the CSI driver. Any credentials required to access (mount) the volume must be made available to the end user consuming the volume.

Reply:
it's possible I'm confusing current phases of volume setup with new CSI-specific concepts. as long as it's possible for a CSI PV to be set up for a user's pod without them ever having PV-related secrets in their namespace, that's the main thing I care about

Reply:
That will depend on what the secret is for. As spec'd, if a volume plugin requires credentials on mount, they must exist in the user's namespace. The assumption there is that a plugin would only require them, if it needs to validate that the user has access to that volume, and therefore it makes sense to  require the secret to be in the user's namespace. For credentials that grant permissions to the entire control plane, those can be injected into the 3rd party volume plugin driver. For provisioner and attach permission, they would work as mentioned above.

I set up a meeting for us to sort this out tomorrow.